## Issues found

Bugs are not prioritised as there is not enough information on the context to
do this.

| ID  | Name                                              | Severity | Test Case (Clickable)                                                                                                                                   | Description                                                                                                                                                                                                                                                         |
|:---:|:--------------------------------------------------|:---------|:--------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|  1  | `order_id` is not uuid                            | medium   | [`test_create_order`](https://github.com/elenakulgavaya/sample-api-tests/blob/main/tests/create_order/test_create_order.py#L12)                         | CreateOrder response contains `order_id` as numeric string.<hr>**Expected** format is UUID                                                                                                                                                                          |
|  2  | `delivery_date` has unexpected format             | medium   | [`test_create_order`](https://github.com/elenakulgavaya/sample-api-tests/blob/main/tests/create_order/test_create_order.py#L12)                         | **Expected** format is `YYYY-MM-DD`                                                                                                                                                                                                                                 |
|  3  | `discount_applied` doesnt affect `total_amount`   | high     | [`test_create_order_with_discount`](https://github.com/elenakulgavaya/sample-api-tests/blob/main/tests/create_order/test_discounts.py#L12)              | When requesting `orders/create` with the discount, the response indicates discount has been applied by the valid value in `discount_applied` field but the `total_amount` doesn't show any effects of it.<hr>**Expected:** discount is reflected in the total price |
|  4  | `note` is not escaped                             | highest  | [`test_check_if_note_is_escaped`](https://github.com/elenakulgavaya/sample-api-tests/blob/main/tests/create_order/test_validation.py#L93)               | The `note` field is saved to db and can be potential place for SQL injection, the value should not be directly parsed to the insert query.<hr>**Expected:** to use query formatting or ORM                                                                          |
|  5  | new order is created on error with `note`         | low      | [`test_check_if_note_is_escaped_teardown`](https://github.com/elenakulgavaya/sample-api-tests/blob/main/tests/create_order/test_validation.py#L93)      | When the result of endpoint is error we are not expecting new record in the system, however in this case the record is created.<hr>**Expected:** no new record on error                                                                                             |
|  6  | `product_id` is not validated                     | medium   | [`test_invalid_product_id`](https://github.com/elenakulgavaya/sample-api-tests/blob/main/tests/create_order/test_validation.py#L27)                     | Any value is acceptable for the product id<hr>**Expected:** Validation error corresponding to `type:integer` requirement                                                                                                                                            |
|  7  | `quantity` is not validated when is number        | medium   | [`test_invalid_quantity_number`](https://github.com/elenakulgavaya/sample-api-tests/blob/main/tests/create_order/test_validation.py#L45)                | Zero, negatives and floats are accepted as quantity<hr>**Expected:** Validation error corresponding to `type: integer` or `minimum: 1` requirements                                                                                                                 |
|  8  | `price_per_unit` can be zero                      | low      | [`test_invalid_price_per_unit`](https://github.com/elenakulgavaya/sample-api-tests/blob/main/tests/create_order/test_validation.py#L65)                 | Minimum acceptable price per unit is 0.01<hr>**Expected:** Validation error corresponding to `minimum: 0.01` requirement                                                                                                                                            |
|  9  | `discount_rate` is not validated                  | medium   | [`test_invalid_discount_rate`](https://github.com/elenakulgavaya/sample-api-tests/blob/main/tests/create_order/test_validation.py#L86)                  | There is no validation for the `discount_rate` and it doesn't seem to be included into any calculation<hr>**Expected:** `type: number`, `minimum: 0`, `maximum: 1`                                                                                                  |
| 10  | `product_id` is not required                      | medium   | [`test_missing_required_field_product_id`](https://github.com/elenakulgavaya/sample-api-tests/blob/main/tests/create_order/test_validation.py#L100)     | `product_id` should be required, when absent is just None at the response<hr>**Expected:** response_code 400, error message for the `product_id` field to be required                                                                                               |
| 11  | `quantity` is not required                        | medium   | [`test_missing_required_field_quantity`](https://github.com/elenakulgavaya/sample-api-tests/blob/main/tests/create_order/test_validation.py#L108)       | `quantity` is expected to be required. Missing the value in request causes internal error `float() argument must be a string or a number, not 'NoneType'`<hr>**Expected:** response_code 400, error message for the `quantity` field to be required                 |
| 12  | `delivery_date` is not required                   | medium   | [`test_missing_required_field_delivery_date`](https://github.com/elenakulgavaya/sample-api-tests/blob/main/tests/create_order/test_validation.py#L116)  | `delvery_date` is expected to be required. Missing the value in request causes internal error `strptime() argument 1 must be str, not None`<hr>**Expected:** response_code 400, error message for the `delivery_date` field to be required                          |
| 13  | `price_per_unit` is not required                  | medium   | [`test_missing_required_field_price_per_unit`](https://github.com/elenakulgavaya/sample-api-tests/blob/main/tests/create_order/test_validation.py#L124) | `price_per_unit` is expected to be required. Missing the value in request causes internal error `float() argument must be a string or a number, not 'NoneType'`<hr>**Expected:** response_code 400, error message for the `price_per_unit` field to be required     |
| 14  | User is authorized by default                     | highest  | [`test_missing_token`](https://github.com/elenakulgavaya/sample-api-tests/blob/main/tests/test_auth.py#L15)                                             | When there is no token provided the user is let to perform create order<hr>**Expected:** response_code 403, error message Forbidden                                                                                                                                 |

[Back to Strategy](strategy.md)

[BACK](../README.md)
